### DISCUSSION TITLE
Button

### DISCUSSION THREAD

**Discussion by bameyrick** [ONSdigital/design-system#1389] [Feedback]
Use this place to discuss the [button component](https://ons-design-system.netlify.com/components/button) in the [ONS Design System](https://ons-design-system.netlify.com)

**Comment by bameyrick** [comment]
The ability to use the button component as anything other than a `<button>` element should be removed as this is bad practice.

**Comment by bameyrick** [comment]
Secondary button font colour is still blue, this should be black (#22222). This appears to be lost during a merge.

**Reply by jrbarnes9** [reply]
Issue #351 resolved

**Comment by bameyrick** [comment]
Small button variant is too tall since underline change has gone in. Its currently 45px tall where as it should be 35px tall.

**Reply by jrbarnes9** [reply]
Issue #364 resolved

**Comment by jrbarnes9** [comment]
I think `btn--small` could do with a bit more left/right padding.

**Reply by jrbarnes9** [reply]
Issue #396 resolved

**Comment by bameyrick** [comment]
I agree... First screenshot is current and second is with horizontal padding set to `0.7em`:

<img width="73" alt="Screenshot 2019-05-24 at 15 56 21" src="https://user-images.githubusercontent.com/8447446/58336979-b4ebca80-7e3c-11e9-98ab-98920359b0be.png">
<img width="77" alt="Screenshot 2019-05-24 at 15 57 06" src="https://user-images.githubusercontent.com/8447446/58336984-b87f5180-7e3c-11e9-94ed-59d4ae0a461d.png">

Fancy chiming in @will-madebymoon ?

**Reply by jrbarnes9** [reply]
Issue #396 resolved

**Comment by jrbarnes9** [comment]
LGTM. I think @will-madebymoon agrees.

**Comment by will-madebymoon** [comment]
I prefer the second one with a little bit wider... 

Sent whilst on the move

**Comment by jrbarnes9** [comment]
Ghost button variant doesn't have a background colour when viewed in a new window: https://ons-design-system.netlify.com/components/button/examples/button-ghost/

**Reply by jrbarnes9** [reply]
Issue #500 resolved

**Comment by bameyrick** [comment]
Pointer events need to be sent to none on `button__inner` to allow click events to be captured correctly

**Reply by jrbarnes9** [reply]
Issue #507 resolved

**Comment by rmccar** [comment]
Need to add a parameter to pass in URL to button component so that it can be used as a link

**Reply by jrbarnes9** [reply]
#740 resolved

**Comment by jrbarnes9** [comment]
Focus state shouldn't change when the element is hovered.

**Reply by jrbarnes9** [reply]
#739 resolved

**Comment by trevorsaint** [comment]
I think we need a button that looks like a link. I also would expect `.btn--link` to be a modifier to do exactly that.

However we are using the modifier `.btn--link` to make links look like buttons. Do we really need to do this? Can’t we just assume that if an `<a>`, `<button>` has a class of `.btn` it will look like a button.

I do feel this is a much better approach, and we can use the modifier to have a `<button>` that looks like a link.

**GDS** do it this way and it does make more sense. 
https://design-system.service.gov.uk/components/button/

**Reply by jrbarnes9** [reply]
#761 raised

**Comment by jrbarnes9** [comment]
Do we need a loader and timer button? This should probably be default behaviour for a `submit` button, as you'd never want user to double-click it, and you'd always need to show user feedback (spinner) if the process takes longer than a certain delay.

**Reply by jrbarnes9** [reply]
>rmccar on 30 Jul 2020:
>It is linked to setting the button type to be a submit button but I suppose we could change the logic a little bit so that it doesn't allow you to have a submit button without setting a submit type?

**Comment by jrbarnes9** [comment]
Macro table has created a button when trying to document the `<button>` tag - missing code markdown markup around it

<img width="591" alt="Screenshot 2021-01-08 at 15 38 43" src="https://user-images.githubusercontent.com/43346934/104033671-bdd31080-51c7-11eb-916b-4db4b653a248.png">

**Reply by jrbarnes9** [reply]
Issue #1279 resolved

**Comment by jrbarnes9** [comment]
Adjust loader button spinner for small variant

![Screenshot 2021-02-23 at 10 36 44](https://user-images.githubusercontent.com/43346934/108832454-7a8c0000-75c3-11eb-9e29-582c11cb2c7d.png)

**Reply by jrbarnes9** [reply]
#1400 resolved

**Comment by jrbarnes9** [comment]
Button shadows are inconsistent between regular and small variants

![Screenshot 2021-05-26 at 10 31 16](https://user-images.githubusercontent.com/43346934/119639314-67314080-be0f-11eb-98dc-aa768213ee7f.png)
![Screenshot 2021-05-26 at 10 31 09](https://user-images.githubusercontent.com/43346934/119639322-68fb0400-be0f-11eb-8bdf-520fc24d41d3.png)

Should all be e.g.

![Screenshot 2021-05-26 at 10 43 27](https://user-images.githubusercontent.com/43346934/119639269-5c76ab80-be0f-11eb-9eb7-2e12118e64c1.png)

**Reply by jrbarnes9** [reply]
#1531 raised

**Comment by jrbarnes9** [comment]
I think the print button icon should be placed before the button label, not after.  The purpose of the icon is to visually convey the purpose of the button, so it can be understood more quickly without needing to read the text label.

e.g.
![Screenshot 2021-08-25 at 12 24 07](https://user-images.githubusercontent.com/43346934/130782609-bc1522a6-4d42-40ff-9f1c-2dfaa303ebf6.png)

Add to icon usage documentation: All icons used to visually represent the purpose of the element should be placed before the text element.  Icons that represent the direction of interaction (e.g. CTA button, Menu button) should be displayed after the text element.

Elements with icons to represent purpose:
- Success panel or table row (Tick)
- Print button (printer) **Needs repositioning*
- Social media link (brand logo)
- Security info panel (lock)
- Blockquote (quote mark)
- Person (head & shoulders)
- Warning panel (exclamation mark)
- Status indicator (dot)

Elements with icons to indication direction of interaction:
- Menu button (down chevron)
- Exit button (door with arrow)
- External link (window with arrow)
- Table column sort (up/down chevrons)

**Reply by jrbarnes9** [reply]
#1656 resolved

**Comment by jrbarnes9** [comment]
We need a "save/download as file" button with icon for use in EQ where users can save a copy of their survey answers as a PDF.

e.g.
<img width="289" alt="Screenshot 2021-08-25 at 12 47 39" src="https://user-images.githubusercontent.com/43346934/130785254-5eb0b8ce-6aee-4a5a-84c6-960f9399acfe.png">

**Reply by jrbarnes9** [reply]
#1655 resolved

**Comment by jrbarnes9** [comment]
Disabled button should use `aria-disabled` attribute so screen readers can discover it's disabled state

https://a11y-101.com/development/aria-disabled

**Reply by armstrongb** [reply]
Also worth a read. https://css-tricks.com/making-disabled-buttons-more-inclusive/

**Comment by jrbarnes9** [comment]
Print button icon not from brand guidelines: https://www.notion.so/ons-design/Iconography-87521a48ce4d4c2b85b71fe455a7ca5e?p=ae78b8e94e84458087210cfd444f2707

![Screenshot 2022-03-24 at 15 04 17](https://user-images.githubusercontent.com/43346934/159946573-861a27de-e48d-45e1-a35d-3be78d713c8a.png)

Should be using: 
![Screenshot 2022-03-24 at 15 04 05](https://user-images.githubusercontent.com/43346934/159946626-2aa0f7b9-fbb0-47d9-a96b-56c022f41f4b.png)

**Reply by jrbarnes9** [reply]
Raised #2085

**Comment by jrbarnes9** [comment]
Disabled button shouldn't have a hover or active state

**Reply by jrbarnes9** [reply]
Raised #2436

**Comment by jrbarnes9** [comment]
Change `buttonContext` macro option description as it references an example that no longer exists

**Reply by jrbarnes9** [reply]
Raised #2512

**Comment by jrbarnes9** [comment]
`variants` missing from Button macro options documentation

**Reply by jrbarnes9** [reply]
Raised #2582

**Comment by Al-Davies** [comment]
**UR Findings**
users clearly seeing these CTA buttons - on census home page (start census), within eq (continue, save and signout) - submit census > no indication of any confusion with this
participants clear they need to click this button in order to submit the form they have completed. Until they click this they know the information won't get sent.
8.4.20
secondary button used on voluntary questions for users to clear answers. Participant noting button, but not necessarily relating it to the question being voluntary - just a way to clear/amend what they've put
